tests/fourier/test_advanced_spectrum.py

```diff
@@ -0,0 +1,442 @@
+        specified by ``encoding_args``."""
+        dev = qml.device("default.qubit", wires=2)
+        z_0 = 2.1
+
+        @qml.qnode(dev)
+        def circuit(x, Y, z=z_0):
+            qml.RX(z * x, wires=0)
+            qml.RY(0.2 * Y[0, 1, 0], wires=0)
+            qml.RY(3 * Y[0, 0, 0], wires=1)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = -1.5
+        Y = np.array([0.2, -1.2, 9.2, -0.2, 1.1, 4, -0.201, 0.8]).reshape((2, 2, 2))
+        z = 1.2
+
+        res = advanced_spectrum(circuit, encoding_args={"x"})(x, Y, z=z)
+        assert res == {"x": {(): [-z, 0.0, z]}}
+
+        res = advanced_spectrum(circuit, encoding_args={"x"})(x, Y)
+        assert res == {"x": {(): [-z_0, 0.0, z_0]}}
+
+        res = advanced_spectrum(circuit, encoding_args={"x": ..., "Y": [(0, 0, 0), (1, 0, 1)]})(
+            x, Y
+        )
+        assert res == {
+            "x": {(): [-z_0, 0.0, z_0]},
+            "Y": {(0, 0, 0): [-3.0, 0.0, 3.0], (1, 0, 1): [0.0]},
+        }
+
+    def test_spectrum_changes_with_qnode_args(self):
+        """Test that the spectrum changes per call if a qnode keyword argument
+        changes the circuit architecture."""
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def circuit(x, last_gate=False):
+            qml.RX(x, wires=0)
+            qml.RX(x, wires=1)
+            if last_gate:
+                qml.RX(x, wires=2)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = 0.9
+        res_true = advanced_spectrum(circuit, argnum=[0])(x, last_gate=True)
+        assert np.allclose(res_true["x"][()], range(-3, 4))
+
+        res_false = advanced_spectrum(circuit, argnum=[0])(x, last_gate=False)
+        assert np.allclose(res_false["x"][()], range(-2, 3))
+
+    def test_multi_par_error(self):
+        """Test that an error is thrown if the spectrum of
+        a multi-parameter gate that cannot be decomposed is requested."""
+        dev = qml.device("default.qubit", wires=3)
+
+        class nondecompRot(qml.Rot):
+            @staticmethod
+            def decomposition(phi, theta, omega, wires):
+                """Pseudo-decomposition: Just return the gate itself."""
+                return [nondecompRot(phi, theta, omega, wires=wires)]
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RX(x, wires=0)
+            nondecompRot(0.2, x, 0.4, wires=1)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with pytest.raises(ValueError, match="Can only consider one-parameter gates"):
+            advanced_spectrum(circuit)(1.5)
+
+    def test_multi_parameter_expansion(self):
+        """Test that a multi-parameter gate is decomposed correctly.
+        And that single-parameter gates are not decomposed."""
+        dev = qml.device("default.qubit", wires=3)
+
+        class _CRX(qml.CRX):
+            @staticmethod
+            def decomposition(theta, wires):
+                raise NotImplementedError()
```

Comment:
wait, how is this gate decomposed, when it looks like there is no decomposition?

Reply:
It is not. And it should not be decomposed, because it is a single-parameter gate! :)